### PR TITLE
Populated 'unexpected' resources while comparing target and live states

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/104

### CODE REVIEW

File: controller/comparator.go [hunk 1]
```diff
@@ -154,17 +173,31 @@ func (ks *KsonnetAppComparator) CompareAppState(
...
-func getChildren(parent *unstructured.Unstructured, objByFullName map[string]*unstructured.Unstructured) ([]v1alpha1.ResourceNode, error) {
+func hasParent(obj *unstructured.Unstructured) bool {
+	// TODO: remove special case after Service and Endpoint get explicit relationship ( https://github.com/kubernetes/kubernetes/issues/28483 )
+	return obj.GetKind() == kubeutil.EndpointsKind || metav1.GetControllerOf(obj) != nil
+}
+
+func isControlledBy(obj *unstructured.Unstructured, parent *unstructured.Unstructured) bool {
+	// TODO: remove special case after Service and Endpoint get explicit relationship ( https://github.com/kubernetes/kubernetes/issues/28483 )
+	if obj.GetKind() == kubeutil.EndpointsKind && parent.GetKind() == kubeutil.ServiceKind {
+		return obj.GetName() == parent.GetName()
```
Comment (jessesuen) [inline controller/comparator.go]: Nice!

File: controller/comparator.go [hunk 2]
```diff
@@ -105,33 +115,42 @@ func (ks *KsonnetAppComparator) CompareAppState(
...
+		if diffResult.Modified {
+			resState.Status = v1alpha1.ComparisonStatusOutOfSync
+		} else {
+			resState.Status = v1alpha1.ComparisonStatusSynced
 		}
-		resState.TargetState = string(targetObjBytes)
+
+		if targetObjs[i] == nil {
+			resState.TargetState = "null"
+			resState.Status = v1alpha1.ComparisonStatusOutOfSync
```
Comment (jessesuen) [inline controller/comparator.go]: I have no problems with the logic, but it takes some time to understand it. Can we add comments about each scenario and why we are considering it out of sync? For example:

1) object is in target, but does not exist in live
2) object is in target and live, but differ
3) object is not defined in target, but exists in the live state

**Review by jessesuen — APPROVED** [review]
Just add some comments for each scenario why we decide something is OutOfSync.

File: controller/comparator.go (hunk 2 above)
Comment (alexmt) [reply controller/comparator.go]: done

**Pull request merged**